src/fft_common.f90

```diff
@@ -75,93 +88,142 @@ subroutine fft_init_arg(pencil)     ! allow to handle Z-pencil input
 end subroutine fft_init_arg
 
 ! Initialise the FFT library to perform arbitrary size transforms
-subroutine fft_init_general(pencil, nx, ny, nz)
+subroutine fft_init_general(pencil, nx, ny, nz, opt_inplace, opt_inplace_r2c, opt_inplace_c2r)
 
    implicit none
 
-   integer, intent(IN) :: pencil
-   integer, intent(IN) :: nx, ny, nz
+   integer, intent(IN) :: pencil, nx, ny, nz
+   logical, intent(in), optional :: opt_inplace, opt_inplace_r2c, opt_inplace_c2r
+
+   ! Only one FFT engine will be used
+   call decomp_2d_fft_set_ngrid(1)
+
+   ! Initialise the FFT engine
+   call decomp_2d_fft_init(pencil, nx, ny, nz, 1, opt_inplace, opt_inplace_r2c, opt_inplace_c2r)
+
+end subroutine fft_init_general
+
+! Initialise the provided FFT grid
+subroutine fft_init_multigrid(pencil, nx, ny, nz, igrid, opt_inplace, opt_inplace_r2c, opt_inplace_c2r)
+
+   implicit none
+
+   integer, intent(in) :: pencil, nx, ny, nz, igrid
+   logical, intent(in), optional :: opt_inplace, opt_inplace_r2c, opt_inplace_c2r
+
+   ! Safety check
+   if (igrid < 1 .or. igrid > n_grid) then
+      call decomp_2d_abort(__FILE__, __LINE__, igrid, "Invalid value for igrid")
+   end if
+
+   ! Initialise the engine
+   call fft_engines(igrid)%init(pencil, nx, ny, nz, opt_inplace, opt_inplace_r2c, opt_inplace_c2r)
+
+end subroutine fft_init_multigrid
 
-   integer :: errorcode
+! Initialise the given FFT engine
+subroutine decomp_2d_fft_engine_init(engine, pencil, nx, ny, nz, opt_inplace, opt_inplace_r2c, opt_inplace_c2r)
+
+   implicit none
+
+   class(decomp_2d_fft_engine), target, intent(inout) :: engine
+   integer, intent(in) :: pencil, nx, ny, nz
+   logical, intent(in), optional :: opt_inplace, opt_inplace_r2c, opt_inplace_c2r
 
-#ifdef PROFILER
    if (decomp_profiler_fft) call decomp_profiler_start("fft_init")
-#endif
 
    ! Safety checks
-   if (initialised) then
-      errorcode = 4
-      call decomp_2d_abort(__FILE__, __LINE__, errorcode, &
+   if (engine%initialised) then
+      call decomp_2d_abort(__FILE__, __LINE__, 4, &
                            'FFT library should only be initialised once')
```

Comment:
Replace error message -> "FFT engine should only be initialised once", could also add the engine ID if possible.

Reply:
Fixed in https://github.com/2decomp-fft/2decomp-fft/pull/302/commits/c22ce37eb50ee668b668c70c6194b4019856a5c1.

Currently, the engine ID is not defined inside the object `decomp_2d_fft_engine`. The ID could be defined when the engine is part of the array `fft_engines`. However, the ID can hardly be defined when the engine is managed in the calling code.

If needed, we could add a ID field using an integer inside the object `decomp_2d_fft_engine`. The ID would be the number of the engine in the array `fft_engine`. Or it would be zero if the engine is managed by the external code. However I am not sure this is very useful.